File: bulk.whois/ruby/bulk_whois_api_sample.rb
require 'json'
require 'net/https'
require 'openssl'
require 'uri'

domains = %w[
  google.com
  example.com
  whoisxmlapi.com
  twitter.com
]

csv_filename = 'whois_records.csv'

########################
# Fill in your details #
########################

def call_api(path, data)
  base_data = {
    apiKey: 'Your bulk whois api key',
    outputFormat: 'json'
  }

  url = 'https://www.whoisxmlapi.com/BulkWhoisLookup/bulkServices' + path

  body = base_data.merge(data).to_json

  http_post(url, body)
end

def create_request(domains)
  data = {
    domains: domains
  }

  response = JSON.parse(call_api('/bulkWhois', data))

  code = response['messageCode']
  message = response['message']

  raise "#{code}: #{message}" if code != 200

  response['requestId']
end

def download(id, file_name)
  data = {
    requestId: id,
    startIndex: 1
  }

  File.write(file_name, call_api('/download', data))
end

def get_records(id, start, max)
  data = {
    requestId: id,
    startIndex: start,
    maxRecords: max
  }

  JSON.parse(call_api('/getRecords', data))
end

def http_post(url, data)
  uri = URI.parse(url)
  http = Net::HTTP.new(uri.host, uri.port)

  # Connect using ssl
  http.use_ssl = true
  http.verify_mode = OpenSSL::SSL::VERIFY_NONE
  request = Net::HTTP::Post.new(uri.request_uri)

  # Set headers
  request.add_field('Content-Type', 'application/json')
  request.add_field('Accept', 'application/json')
  request.body = data

  http.request(request).body
end

puts 'Requesting bulk whois lookup...'

id = create_request(domains)

puts "Request created: #{id}"

puts 'Waiting for processing...'

loop do
  response = get_records(id, domains.count + 1, 0)
  records_left = response['recordsLeft']

  puts "Records left: #{records_left}"

  break if records_left.zero?

  sleep 3
end

puts 'Downloading results...'

download(id, csv_filename)

puts "Saved to #{csv_filename}"

File: bulk.whois/node/bulk_whois_api_sample.js
var fs = require('fs'), https = require('https');

var dom = ['google.com', 'youtube.com', 'facebook.com', 'whoisxmlapi.com'];
var srv = 'www.whoisxmlapi.com', url = '/BulkWhoisLookup/bulkServices/';
var apiKey = 'Your bulk whois api key';
var file = 'bulk.csv';

function api(path, data, callback)
{
    var baseData = {
        apiKey: apiKey,
        outputFormat: 'json'
    };

    var body = '';
    var opts = {
        host: srv,
        path: url + path,
        method:'POST',
        headers: {'Content-Type': 'application/json'}
    };

    https.request(opts, function(response) {
        response.on('data', function(chunk) {
            body += chunk;
        });
        response.on('end', function() {
            callback(body);
        });
    }).end(JSON.stringify(Object.assign({}, baseData, data)));
}

// This will save whois record info for all domains as 'bulk.csv' in current
// directory.

api('bulkWhois', {domains:dom}, function(body) {
    var id = {
        requestId: JSON.parse(body).requestId,
        startIndex: dom.length + 1
    };
    var timer = setInterval(function() {
        api('getRecords', Object.assign({},id,{maxRecords:0}),function(body) {
            if (JSON.parse(body).recordsLeft) {
                return;
            }
            clearInterval(timer);
            if (typeof JSON.parse(body).recordsLeft === 'undefined') {
                return;
            }
            api('download', id, function(csv) {
                fs.writeFile(file,csv);
            });
        });
    }, 3000);
});
